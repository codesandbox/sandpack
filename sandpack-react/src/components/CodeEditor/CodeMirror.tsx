import { useClasser } from "@code-hike/classer";
import { closeBrackets, closeBracketsKeymap } from "@codemirror/closebrackets";
import {
  defaultKeymap,
  indentLess,
  deleteGroupBackward,
  insertTab,
} from "@codemirror/commands";
import { commentKeymap } from "@codemirror/comment";
import { lineNumbers } from "@codemirror/gutter";
import { defaultHighlightStyle } from "@codemirror/highlight";
import { history, historyKeymap } from "@codemirror/history";
import { bracketMatching } from "@codemirror/matchbrackets";
import { EditorState } from "@codemirror/state";
import type { Annotation, Extension } from "@codemirror/state";
import {
  highlightSpecialChars,
  highlightActiveLine,
  keymap,
  EditorView,
} from "@codemirror/view";
import type { KeyBinding } from "@codemirror/view";
import useIntersectionObserver from "@react-hook/intersection-observer";
import * as React from "react";

import { useSandpack } from "../../hooks/useSandpack";
import { useSandpackTheme } from "../../hooks/useSandpackTheme";
import type {
  EditorState as SandpackEditorState,
  SandpackInitMode,
} from "../../types";
import { getFileName, generateRandomId } from "../../utils/stringUtils";

import { highlightDecorators } from "./highlightDecorators";
import { highlightInlineError } from "./highlightInlineError";
import {
  getCodeMirrorLanguage,
  getEditorTheme,
  getSyntaxHighlight,
  useCombinedRefs,
} from "./utils";

export type Decorators = Array<{
  className?: string;
  line: number;
  startColumn?: number;
  endColumn?: number;
  elementAttributes?: Record<string, string>;
}>;

interface CodeMirrorProps {
  code: string;
  filePath?: string;
  fileType?:
    | "js"
    | "jsx"
    | "ts"
    | "tsx"
    | "css"
    | "scss"
    | "less"
    | "html"
    | "vue";
  onCodeUpdate?: (newCode: string) => void;
  showLineNumbers?: boolean;
  showInlineErrors?: boolean;
  wrapContent?: boolean;
  editorState?: SandpackEditorState;
  /**
   * This disables editing of content by the user in all files.
   */
  readOnly?: boolean;
  /**
   * Controls the visibility of Read-only label, which will only
   * appears when `readOnly` is `true`
   */
  showReadOnly?: boolean;
  decorators?: Decorators;
  initMode: SandpackInitMode;
  id?: string;
  extensions?: Extension[];
  extensionsKeymap?: Array<readonly KeyBinding[]>;
}

export interface CodeMirrorRef {
  getCodemirror: () => EditorView | undefined;
}

/**
 * @category Components
 */
export const CodeMirror = React.forwardRef<CodeMirrorRef, CodeMirrorProps>(
  (
    {
      code,
      filePath,
      fileType,
      onCodeUpdate,
      showLineNumbers = false,
      showInlineErrors = false,
      wrapContent = false,
      editorState = "pristine",
      readOnly = false,
      showReadOnly = true,
      decorators,
      initMode = "lazy",
      id,
      extensions = [],
      extensionsKeymap = [],
    },
    ref
  ) => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const wrapper = React.useRef<any | HTMLElement>(null);
    const cmView = React.useRef<EditorView>();
    const { theme, themeId } = useSandpackTheme();
    const [internalCode, setInternalCode] = React.useState<string>(code);
    const [shouldInitEditor, setShouldInitEditor] = React.useState(
      initMode === "immediate"
    );

    const c = useClasser("sp");
    const { listen } = useSandpack();
    const ariaId = React.useRef<string>(id ?? generateRandomId());

    const { isIntersecting } = useIntersectionObserver(wrapper, {
      rootMargin: "600px 0px",
      threshold: 0.2,
    });

    React.useImperativeHandle(ref, () => ({
      getCodemirror: (): EditorView | undefined => cmView.current,
    }));

    React.useEffect(() => {
      const mode = initMode === "lazy" || initMode === "user-visible";

      if (mode && isIntersecting) {
        setShouldInitEditor(true);
      }
    }, [initMode, isIntersecting]);

    React.useEffect(() => {
      if (!wrapper.current || !shouldInitEditor) return;

      /**
       * TODO: replace this time out to something more efficient
       * waiting for "postTask scheduler" API be ready
       */
      const timer = setTimeout(function delayCodeEditorInit() {
        const langSupport = getCodeMirrorLanguage(filePath, fileType);

        const customCommandsKeymap: KeyBinding[] = [
          {
            key: "Tab",
            run: insertTab,
          },
          {
            key: "Shift-Tab",
            run: indentLess,
          },
          {
            key: "Escape",
            run: (): boolean => {
              if (readOnly) return true;

              if (wrapper.current) {
                wrapper.current.focus();
              }

              return true;
            },
          },
          {
            key: "mod-Backspace",
            run: deleteGroupBackward,
          },
        ];

        const extensionList = [
          highlightSpecialChars(),
          history(),
          closeBrackets(),

          keymap.of([
            ...closeBracketsKeymap,
            ...defaultKeymap,
            ...historyKeymap,
            ...commentKeymap,
            ...customCommandsKeymap,
            ...extensionsKeymap,
          ] as KeyBinding[]),
          langSupport,

          defaultHighlightStyle.fallback,

          getEditorTheme(theme),
          getSyntaxHighlight(theme),
          ...extensions,
        ];

        if (readOnly) {
          extensionList.push(EditorView.editable.of(false));
        } else {
          extensionList.push(bracketMatching());
          extensionList.push(highlightActiveLine());
        }

        if (decorators) {
          extensionList.push(highlightDecorators(decorators));
        }

        if (wrapContent) {
          extensionList.push(EditorView.lineWrapping);
        }

        if (showLineNumbers) {
          extensionList.push(lineNumbers());
        }

        if (showInlineErrors) {
          extensionList.push(highlightInlineError());
        }

        const startState = EditorState.create({
          doc: code,
          extensions: extensionList,
        });

        const parentDiv = wrapper.current;

        const view = new EditorView({
          state: startState,
          parent: parentDiv,
          dispatch: (tr): void => {
            view.update([tr]);

            if (tr.docChanged) {
              const newCode = tr.newDoc.sliceString(0, tr.newDoc.length);
              setInternalCode(newCode);
              onCodeUpdate?.(newCode);
            }
          },
        });

        view.contentDOM.setAttribute("data-gramm", "false");

        if (!readOnly) {
          view.contentDOM.setAttribute("tabIndex", "-1");
          view.contentDOM.setAttribute(
            "aria-describedby",
            `exit-instructions-${ariaId.current}`
          );
        }

        cmView.current = view;
      }, 0);

      return (): void => {
        cmView.current?.destroy();

        clearTimeout(timer);
      };

      // TODO: Would be nice to reconfigure the editor when these change, instead of recreating with all the extensions from scratch
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [shouldInitEditor, showLineNumbers, wrapContent, themeId, decorators]);

    React.useEffect(() => {
      // When the user clicks on a tab button on a larger screen
      // Avoid autofocus on mobile as it leads to a bad experience and an unexpected layout shift
      if (
        cmView.current &&
        editorState === "dirty" &&
        window.matchMedia("(min-width: 768px)").matches
      ) {
        cmView.current.contentDOM.focus();
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    // Update editor when code passed as prop from outside sandpack changes
    React.useEffect(() => {
      if (cmView.current && code !== internalCode) {
        const view = cmView.current;
        view.dispatch({
          changes: { from: 0, to: view.state.doc.length, insert: code },
        });
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [code]);

    React.useEffect(
      function messageToInlineError() {
        if (!showInlineErrors) return;

        const unsubscribe = listen((message) => {
          const view = cmView.current;

          if (message.type === "success") {
            view?.dispatch({
              // Pass message to clean up inline error
              annotations: [
                {
                  type: "clean-error",
                  value: null,
                } as unknown as Annotation<unknown>,
              ],

              // Trigger a doc change to remove inline error
              changes: {
                from: 0,
                to: view.state.doc.length,
                insert: view.state.doc,
              },
              selection: view.state.selection,
            });
          }

          if (
            message.type === "action" &&
            message.action === "show-error" &&
            "line" in message
          ) {
            view?.dispatch({
              annotations: [
                {
                  type: "error",
                  value: message.line,
                } as unknown as Annotation<unknown>,
              ],
            });
          }
        });

        return (): void => unsubscribe();
      },
      [listen, showInlineErrors]
    );

    const handleContainerKeyDown = (evt: React.KeyboardEvent): void => {
      if (evt.key === "Enter" && cmView.current) {
        evt.preventDefault();
        cmView.current.contentDOM.focus();
      }
    };

    const combinedRef = useCombinedRefs(wrapper, ref);

    if (readOnly) {
      return (
        <pre ref={combinedRef} className={c("cm", editorState)} translate="no">
          {!shouldInitEditor && (
            <code className={c("pre-placeholder")}>{code}</code>
          )}

          {readOnly && showReadOnly && (
            <span className={c("read-only")}>Read-only</span>
          )}
        </pre>
      );
    }

    return (
      /* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
      /* eslint-disable jsx-a11y/no-noninteractive-tabindex */
      <div
        ref={combinedRef}
        aria-describedby={`enter-instructions-${ariaId.current}`}
        aria-label={
          filePath ? `Code Editor for ${getFileName(filePath)}` : `Code Editor`
        }
        className={c("cm", editorState)}
        onKeyDown={handleContainerKeyDown}
        role="group"
        tabIndex={0}
        translate="no"
      >
        {!shouldInitEditor && (
          <pre
            className={c("pre-placeholder")}
            style={{
              marginLeft: showLineNumbers ? 28 : 0, // gutter line offset
            }}
          >
            {code}
          </pre>
        )}

        <>
          <p
            id={`enter-instructions-${ariaId.current}`}
            style={{ display: "none" }}
          >
            To enter the code editing mode, press Enter. To exit the edit mode,
            press Escape
          </p>
          <p
            id={`exit-instructions-${ariaId.current}`}
            style={{ display: "none" }}
          >
            You are editing the code. To exit the edit mode, press Escape
          </p>
        </>
      </div>
    );
  }
);
