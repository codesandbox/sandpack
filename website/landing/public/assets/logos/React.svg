<svg width="100%" height="100%" viewBox="0 0 147 39" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path fill-rule="evenodd" clip-rule="evenodd" d="M23.822 34.89c2.763 2.335 5.47 3.608 7.583 3.608a3.465 3.465 0 0 0 2.058-.584c1.137-.69 1.95-2.015 2.329-3.925.325-1.75.27-3.978-.217-6.312a40.81 40.81 0 0 0-.258-1.036c.285-.076.57-.155.854-.237 5.091-1.644 8.233-4.35 8.233-6.948 0-1.326-.759-2.6-2.22-3.82-1.355-1.113-3.25-2.12-5.525-2.863a27.73 27.73 0 0 0-1.352-.427c.075-.281.146-.564.214-.847 1.083-5.145.325-9.123-2.004-10.449-1.137-.636-2.708-.636-4.495 0-1.625.584-3.466 1.698-5.254 3.236a44.98 44.98 0 0 0-1.076 1.002c-.27-.266-.543-.53-.82-.79-4.007-3.448-7.907-4.827-10.236-3.5-1.137.636-1.896 1.962-2.22 3.818-.326 1.698-.271 3.82.216 6.1.092.474.2.958.324 1.45-.526.144-1.05.297-1.57.46C3.62 14.47.804 16.91.804 19.456c0 1.326.758 2.652 2.274 3.925a18.22 18.22 0 0 0 5.687 2.97c.418.126.836.245 1.257.358-.14.532-.27 1.067-.39 1.605-.92 4.826-.162 8.433 2.058 9.706.542.319 1.246.478 2.058.478.759 0 1.571-.16 2.654-.425 1.733-.583 3.629-1.75 5.47-3.341.295-.274.586-.551.872-.832.354.336.713.666 1.078.99Zm.204-2.3A44.163 44.163 0 0 0 27.788 28a46.51 46.51 0 0 0 5.675-.904c.095.352.185.705.27 1.06.975 4.667.054 7.584-1.192 8.274-1.679.955-4.928-.796-7.474-2.917a25.303 25.303 0 0 1-1.043-.922Zm10.785-7.71c.242-.073.478-.148.71-.226 4.82-1.538 6.987-3.766 6.987-5.198 0-.69-.542-1.591-1.57-2.387a16.207 16.207 0 0 0-4.93-2.546 44.048 44.048 0 0 0-1.224-.354 43.387 43.387 0 0 1-2.061 5.352 41.376 41.376 0 0 1 2.088 5.36Zm-3.197-3.133a40.642 40.642 0 0 1 1.343 3.62c-1.2.273-2.487.492-3.84.657a43.77 43.77 0 0 0 1.259-2.113 43.81 43.81 0 0 0 1.238-2.164Zm-1.063-2.294a41.763 41.763 0 0 1-1.908 3.45 57.117 57.117 0 0 1-2.12 3.367c-1.298.09-2.637.134-4 .134-1.283.034-2.564.01-3.84-.072a49.484 49.484 0 0 1-2.172-3.376 49.387 49.387 0 0 1-1.915-3.504 42.118 42.118 0 0 1 1.86-3.39 44.357 44.357 0 0 1 2.115-3.302 54.437 54.437 0 0 1 4.006-.146 44.222 44.222 0 0 1 3.976.127 41.8 41.8 0 0 1 2.09 3.32 40.89 40.89 0 0 1 1.908 3.392Zm1.01-2.209c.49-1.163.928-2.348 1.312-3.551a44.179 44.179 0 0 0-3.874-.694c.44.673.863 1.358 1.269 2.054.453.718.884 1.449 1.292 2.191Zm-3.964-6.27c2.047.182 4.006.479 5.813.886.075-.278.147-.558.216-.838.468-1.891.578-3.851.325-5.781-.217-1.38-.705-2.334-1.409-2.705-.65-.372-1.679-.319-2.924.106a16.843 16.843 0 0 0-4.712 2.917c-.343.314-.681.634-1.014.959a46.982 46.982 0 0 1 3.705 4.455Zm-5.022-3.116a42.213 42.213 0 0 1 2.549 2.952 57.883 57.883 0 0 0-2.493-.053c-.865-.012-1.73 0-2.592.038a44.73 44.73 0 0 1 2.536-2.937ZM21.36 6.626a42.347 42.347 0 0 0-.732-.696 16.62 16.62 0 0 0-4.983-3.13c-1.3-.477-2.437-.53-3.087-.158-.65.37-1.084 1.273-1.354 2.546a15.59 15.59 0 0 0 .216 5.41c.089.432.19.874.307 1.323a45.707 45.707 0 0 1 5.995-.968 44.662 44.662 0 0 1 3.638-4.327Zm-10.912 7.477c-.526.148-1.033.305-1.52.473-3.14 1.06-6.336 3.023-6.336 4.88 0 .742.596 1.644 1.625 2.546 1.246 1.008 2.979 1.962 5.091 2.652.4.12.802.233 1.205.34a42.838 42.838 0 0 1 2.092-5.432 42.808 42.808 0 0 1-2.157-5.459Zm3.214 3.341a39.724 39.724 0 0 1-1.433-3.784 42.424 42.424 0 0 1 4.084-.682c-.48.73-.937 1.476-1.372 2.234a43.162 43.162 0 0 0-1.279 2.232Zm-.042 4.148a41.863 41.863 0 0 0-1.4 3.82c1.334.296 2.68.528 4.035.694a43.254 43.254 0 0 1-1.368-2.248 54.222 54.222 0 0 1-1.267-2.266Zm3.98 6.458a44.16 44.16 0 0 1-5.876-.92c-.128.48-.247.963-.358 1.449-.596 3.182-.487 6.842 1.192 7.797 1.245.69 4.278.053 7.961-3.13.273-.235.548-.483.823-.742a43.583 43.583 0 0 1-3.743-4.454Zm2.462.148c.84 1.08 1.704 2.09 2.57 3.004a38.186 38.186 0 0 0 2.524-3.025c-.858.034-1.718.044-2.578.03-.84.021-1.678.018-2.516-.009Zm6.415-8.742c0 2.109-1.746 3.819-3.9 3.819-2.153 0-3.899-1.71-3.899-3.82 0-2.108 1.746-3.818 3.9-3.818 2.153 0 3.9 1.71 3.9 3.819Zm39.7 13.472-7.095-10.873h-5.578v10.873h-3.467V5.612h11.212c5.091 0 8.72 3.182 8.72 8.221 0 4.933-3.412 7.638-7.258 7.957l7.529 11.138h-4.062Zm.217-19.095c0-3.076-2.275-5.198-5.578-5.198h-7.312v10.45h7.312c3.303 0 5.578-2.175 5.578-5.252Zm16.899-1.167c-5.796 0-9.966 4.615-9.966 10.343 0 6.206 4.333 10.397 10.29 10.45 1.476.051 2.947-.19 4.324-.71a10.97 10.97 0 0 0 3.692-2.314l-1.516-2.015c-1.57 1.591-3.954 2.493-6.229 2.493-4.332 0-6.986-3.13-7.257-6.843h16.357v-.795c0-5.94-3.575-10.609-9.695-10.609ZM76.63 21.79c.162-3.024 2.275-6.578 6.608-6.578 4.603 0 6.607 3.607 6.661 6.578H76.63Zm32.822 11.138v-2.227a8.73 8.73 0 0 1-3.05 2.098 8.881 8.881 0 0 1-3.666.66c-3.413 0-7.096-2.228-7.096-6.577 0-4.403 3.63-6.524 7.096-6.524 2.762 0 5.037.848 6.716 2.705v-3.554c0-2.652-2.167-4.19-5.092-4.19-2.437 0-4.387.848-6.174 2.758l-1.462-2.122c2.166-2.228 4.766-3.289 8.07-3.289 4.279 0 7.799 1.91 7.799 6.684v13.578h-3.141Zm0-7.903c-1.192-1.644-3.358-2.44-5.579-2.44-2.925 0-4.983 1.804-4.983 4.35 0 2.493 2.058 4.296 4.983 4.296 2.221 0 4.333-.849 5.579-2.493v-3.713Zm16.194-12.359c-6.012 0-10.074 4.509-10.074 10.343 0 5.888 4.062 10.397 10.074 10.45a8.874 8.874 0 0 0 4.15-.815 8.675 8.675 0 0 0 3.27-2.633l-2.112-1.91c-1.354 1.804-3.033 2.6-5.146 2.6-4.332 0-6.986-3.236-6.986-7.639 0-4.402 2.654-7.637 6.986-7.637a6.06 6.06 0 0 1 2.917.646 5.926 5.926 0 0 1 2.229 1.953l2.112-1.91c-1.571-1.962-3.737-3.448-7.42-3.448Zm12.457 3.183V28.79c0 2.97 1.462 4.615 4.441 4.668a5.3 5.3 0 0 0 2.026-.265 5.232 5.232 0 0 0 1.766-1.008l-.921-2.281c-.379.424-1.246.796-2.112.796-1.409 0-2.059-1.061-2.059-2.547V15.85h4.117v-2.705h-4.117v-5.41h-3.141v5.41h-3.358v2.705h3.358Z" fill="#fff" />
</svg>