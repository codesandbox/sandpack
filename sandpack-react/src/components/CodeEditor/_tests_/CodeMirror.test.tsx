import React from "react";
import { create, act } from "react-test-renderer";

import { SandpackProvider, SandpackThemeProvider } from "../../..";
import { CodeMirror } from "../CodeMirror";

beforeEach(() => {
  jest.useFakeTimers();
});

const Provider: React.FC = ({ children }) => {
  return (
    <SandpackProvider>
      <SandpackThemeProvider>{children}</SandpackThemeProvider>
    </SandpackProvider>
  );
};

describe("CodeMirror", () => {
  test("SyntaxHighlight: HTML", () => {
    let component;

    act(() => {
      component = create(
        <Provider>
          <CodeMirror />
        </Provider>
      );
    });

    act(() => {
      component.update(
        <Provider>
          <CodeMirror />
        </Provider>
      );
    });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
