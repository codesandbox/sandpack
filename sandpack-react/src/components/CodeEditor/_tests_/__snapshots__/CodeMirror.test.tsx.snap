// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeMirror SyntaxHighlight: HTML 1`] = `
<div
  className="sp-wrapper sp-default"
>
  <div
    aria-describedby="enter-instructions-3061"
    aria-label="Code Editor"
    className="sp-cm sp-pristine"
    onKeyDown={[Function]}
    role="group"
    tabIndex={0}
    translate="no"
  >
    <pre
      className="sp-pre-placeholder"
      style={
        Object {
          "marginLeft": 0,
        }
      }
    />
    <p
      id="enter-instructions-3061"
      style={
        Object {
          "display": "none",
        }
      }
    >
      To enter the code editing mode, press Enter. To exit the edit mode, press Escape
    </p>
    <p
      id="exit-instructions-3061"
      style={
        Object {
          "display": "none",
        }
      }
    >
      You are editing the code. To exit the edit mode, press Escape
    </p>
  </div>
</div>
`;
